File: proto/Messages.cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

namespace HfEngine.Protocol
{
    public static class ProtocolConstants
    {
        public const ushort FrameMagic = 0x534D;

        public const byte MsgTypeNewOrder = 0x01;
        public const byte MsgTypeCancelOrder = 0x02;
        public const byte MsgTypeOrderAck = 0x03;
        public const byte MsgTypeOrderExecuted = 0x04;
        public const byte MsgTypeOrderBookL2Delta = 0x05;

        public const byte FlagSideBuy = 0x00;
        public const byte FlagSideSell = 0x01;
        public const byte FlagIOC = 0x02;
        public const byte FlagPostOnly = 0x04;

        public const byte AckStatusAccepted = 0x00;
        public const byte AckStatusRejected = 0x01;
        public const byte AckStatusCanceled = 0x02;

        public const int FrameHeaderSize = 8;
        public const int NewOrderPayloadSize = 32;
        public const int CancelOrderPayloadSize = 16;
        public const int OrderAckPayloadSize = 24;
        public const int OrderExecutedSize = 32;
        public const int OrderBookL2DeltaSize = 24;
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public readonly struct FrameHeader
    {
        public readonly ushort Magic;
        public readonly byte MsgType;
        public readonly byte Flags;
        public readonly ushort PayloadLen;
        public readonly ushort Reserved;

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public FrameHeader(byte msgType, byte flags, ushort payloadLen)
        {
            Magic = ProtocolConstants.FrameMagic;
            MsgType = msgType;
            Flags = flags;
            PayloadLen = payloadLen;
            Reserved = 0;
        }

        public bool IsValid => Magic == ProtocolConstants.FrameMagic;
        public bool IsBuy => (Flags & ProtocolConstants.FlagSideSell) == 0;
        public bool IsSell => (Flags & ProtocolConstants.FlagSideSell) != 0;
        public bool IsIOC => (Flags & ProtocolConstants.FlagIOC) != 0;
        public bool IsPostOnly => (Flags & ProtocolConstants.FlagPostOnly) != 0;
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public readonly struct NewOrderPayload
    {
        public readonly ulong OrderId;
        public readonly ulong PlayerOrTraderId;
        public readonly long Price;
        public readonly uint Quantity;
        public readonly uint AssetId;

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public NewOrderPayload(ulong orderId, ulong playerOrTraderId, long price, uint quantity, uint assetId)
        {
            OrderId = orderId;
            PlayerOrTraderId = playerOrTraderId;
            Price = price;
            Quantity = quantity;
            AssetId = assetId;
        }
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public readonly struct CancelOrderPayload
    {
        public readonly ulong OrderId;
        public readonly ulong PlayerOrTraderId;

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public CancelOrderPayload(ulong orderId, ulong playerOrTraderId)
        {
            OrderId = orderId;
            PlayerOrTraderId = playerOrTraderId;
        }
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public unsafe struct OrderAckPayload
    {
        public ulong OrderId;
        public ulong PlayerOrTraderId;
        public byte Status;
        public fixed byte Reserved[7];

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public OrderAckPayload(ulong orderId, ulong playerOrTraderId, byte status)
        {
            OrderId = orderId;
            PlayerOrTraderId = playerOrTraderId;
            Status = status;
        }
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public readonly struct OrderExecutedPayload
    {
        public readonly ulong TradeId;
        public readonly ulong MakerOrderId;
        public readonly ulong TakerOrderId;
        public readonly long ExecutionPrice;

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public OrderExecutedPayload(ulong tradeId, ulong makerOrderId, ulong takerOrderId, long executionPrice)
        {
            TradeId = tradeId;
            MakerOrderId = makerOrderId;
            TakerOrderId = takerOrderId;
            ExecutionPrice = executionPrice;
        }
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public unsafe struct OrderBookL2DeltaPayload
    {
        public uint AssetId;
        public uint Reserved1;
        public long Price;
        public uint NewQuantity;
        public byte Side;
        public fixed byte Reserved2[3];

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public OrderBookL2DeltaPayload(uint assetId, long price, uint newQuantity, byte side)
        {
            AssetId = assetId;
            Reserved1 = 0;
            Price = price;
            NewQuantity = newQuantity;
            Side = side;
        }
    }
}

File: proto/messages.go
package proto

import (
	"encoding/binary"
	"errors"
	"unsafe"
)

const (
	FrameMagic uint16 = 0x534D // "SM" in Little-Endian

	MsgTypeNewOrder         uint8 = 0x01
	MsgTypeCancelOrder      uint8 = 0x02
	MsgTypeOrderAck         uint8 = 0x03
	MsgTypeOrderExecuted    uint8 = 0x04
	MsgTypeOrderBookL2Delta uint8 = 0x05

	FlagSideBuy  uint8 = 0x00
	FlagSideSell uint8 = 0x01 // Bit 0
	FlagIOC      uint8 = 0x02 // Bit 1
	FlagPostOnly uint8 = 0x04 // Bit 2

	AckStatusAccepted uint8 = 0x00
	AckStatusRejected uint8 = 0x01
	AckStatusCanceled uint8 = 0x02

	FrameHeaderSize        = 8
	NewOrderPayloadSize    = 32
	CancelOrderPayloadSize = 16
	OrderAckPayloadSize    = 24
	OrderExecutedSize      = 32
	OrderBookL2DeltaSize   = 24
)

var (
	ErrInvalidMagic       = errors.New("invalid magic number in frame header")
	ErrBufferTooShort     = errors.New("buffer too short for expected frame payload")
	ErrUnknownMessageType = errors.New("unknown message type")
)

type FrameHeader struct {
	Magic      uint16
	MsgType    uint8
	Flags      uint8
	PayloadLen uint16
	Reserved   uint16
}

type NewOrderPayload struct {
	OrderId          uint64
	PlayerOrTraderId uint64
	Price            int64
	Quantity         uint32
	AssetId          uint32
}

type CancelOrderPayload struct {
	OrderId          uint64
	PlayerOrTraderId uint64
}

type OrderAckPayload struct {
	OrderId          uint64
	PlayerOrTraderId uint64
	Status           uint8
	Reserved         [7]byte
}

type OrderExecutedPayload struct {
	TradeId        uint64
	MakerOrderId   uint64
	TakerOrderId   uint64
	ExecutionPrice int64
}

type OrderBookL2DeltaPayload struct {
	AssetId     uint32
	Reserved1   uint32
	Price       int64
	NewQuantity uint32
	Side        uint8
	Reserved2   [3]byte
}

func CastFrameHeader(b []byte) *FrameHeader {
	if len(b) < FrameHeaderSize {
		return nil
	}
	return (*FrameHeader)(unsafe.Pointer(&b[0]))
}

func CastNewOrderPayload(b []byte) *NewOrderPayload {
	if len(b) < NewOrderPayloadSize {
		return nil
	}
	return (*NewOrderPayload)(unsafe.Pointer(&b[0]))
}

func CastCancelOrderPayload(b []byte) *CancelOrderPayload {
	if len(b) < CancelOrderPayloadSize {
		return nil
	}
	return (*CancelOrderPayload)(unsafe.Pointer(&b[0]))
}

func CastOrderAckPayload(b []byte) *OrderAckPayload {
	if len(b) < OrderAckPayloadSize {
		return nil
	}
	return (*OrderAckPayload)(unsafe.Pointer(&b[0]))
}

func CastOrderExecutedPayload(b []byte) *OrderExecutedPayload {
	if len(b) < OrderExecutedSize {
		return nil
	}
	return (*OrderExecutedPayload)(unsafe.Pointer(&b[0]))
}

func CastOrderBookL2DeltaPayload(b []byte) *OrderBookL2DeltaPayload {
	if len(b) < OrderBookL2DeltaSize {
		return nil
	}
	return (*OrderBookL2DeltaPayload)(unsafe.Pointer(&b[0]))
}

func EncodeHeader(dst []byte, msgType uint8, flags uint8, payloadLen uint16) {
	binary.LittleEndian.PutUint16(dst[0:2], FrameMagic)
	dst[2] = msgType
	dst[3] = flags
	binary.LittleEndian.PutUint16(dst[4:6], payloadLen)
	binary.LittleEndian.PutUint16(dst[6:8], 0)
}
